### DISCUSSION TITLE
Onboarding Atom users to Pulsar

### DISCUSSION THREAD

**Discussion by Daeraxa** [pulsar-edit/.github#63] [Core Application & Packages]
# Onboarding Atom users to Pulsar
This is so we can discuss what might be needed in order to help onboard Atom users to Pulsar so that the change is as painless as possible, the last thing we need is to make it hard for Atom users to adopt Pulsar which might make them rethink the entire thing and go to another editor entirely. 
We should be able to convert any config files, copy preferences, install packages etc. so that when Pulsar is opened it should be (as near as possible) functionally identical to their old Atom installation.

Once we agree on some/any steps we can look at creating a project to track the progress of this.

## `.atom` -> `.pulsar`
- Pulsar should be able to either automatically find and suggest (but not force) a migration from the old `.atom` dir to `.pulsar` and/or allow a user to manually trigger the migration.

## `init.coffee` -> `init.js`
- Pulsar should be able to automatically convert the init file from coffeescript to javascript (https://github.com/pulsar-edit/pulsar/pull/112). This should be "triggerable" outside of the above migration tool to those who are maybe doing a new install but keep a "dotfiles" repo.

## `.cson` files
- If the decision is to move away from `cson` files (such as `keymap.cson`) then we will need something similar to the above to convert those to the new format. Personally I see [json5](https://json5.org/) as the strongest contender here so we may need something like the `.coffee->.js` automatic conversion.

## Documentation
- We should have a section for those who are coming from Atom to help understand any differences that they need to get used to (for example `apm` -> `pulsar -p`)

**Comment by confused-Techie** [comment]
As for moving the config file, I think a smart way would be to give a couple options. 
* Copy the config to `.pulsar` and leave `.atom` alone. - Just so they still have the old Atom config.
* Replace the `.atom` config for `.pulsar`
* Create a fresh new `.pulsar` config
* See how to do this yourself - Then this could link to a doc on our website about how to do this manually, and what this actually means. Like what settings are actually transferred over and so on, just so that the users that really care can be hyper aware of what this means for their Pulsar and Atom installation.


But beyond this, should we have these pop up as notifications?
Or I think this should be a new pane in the `welcome` package. Maybe we add something to the config or some way to identify first time users, or really someone that hasn't dismissed this specific welcome message. And that could pop up the whole ingest workflow.

Additionally beyond that, if we do decide to use the `welcome` package to display these views, I think having a page showing the changelog can help users that haven't been following our process see what changing to Pulsar does for them. I would want it easy for an end user to see that Pulsar does have progress and advantages over old Atom. Just because we know it may be slightly buggy, but hopefully this can get some good will with them.

**Comment by confused-Techie** [comment]
Also I posted this on Discord, since not everything talked about is relevant here, but thought I should probably copy/paste it here, just for easy reference later on:

---

Also I just mentioned this on discussion #63, but what would we think about the following for the upcoming sunset.

Use the `welcome` package to display a few more things for us.
* A whole page dedicated to the migration of Atom settings to Pulsar. I won't touch to much on that here, but there's more details on the discussion.
* A Changelog, showing the changes we have made AND why they are good. I feel like this can make users more forgiving if/when they encounter issues that didn't exist on Atom.
* Possibly a new, or heavily modify the normal welcome page. Not just to brag about our features, but to take a moment to get users familiar with who we really are. Talking about us taking the reigns from atom and such.

What do we think?

**Reply by GuilleW** [reply]
We can add a "Migrate from Atom" button to welcome package.
And copy the config with translate from cson to json5 ?

**Comment by sertonix** [comment]
It was not my idea but I think it should be here:
Maybe add support for [`.json5`](https://www.npmjs.com/package/json5) files besides `.json` and `.cson`. This means we could also decaffeinate the `config.cson` etc. but don't restrict to normal json.

I have created a [small tool](https://github.com/sertonix/cson-to-json5) that is able to convert cson to json5 and keeps comments. Since cson is missing specifications it might not always work.

**Reply by Daeraxa** [reply]
> The conversion might not be simple but maybe somebody could write a program that converts the format and keeps comments

I think that would be essential, especially people who might have complex and extensively commented cson files.

**Comment by humanplayer2** [comment]
As a current Atom user thinking about moving to a new editor, I would appreciate the "See how to do this yourself" option shown on the [Pulsar-edit README](https://github.com/pulsar-edit/.github/tree/main/profile/README.md).

Perhaps that text could also be used to balance expectations about what packages will and will not work with Pulsar, i.e., what packages it will and will not be possible to migrate.

**Reply by Daeraxa** [reply]
In theory most of the packages should work as they did in Atom, any that don't need to be reported so we can see if it is something we did and if so, if we can fix it.

As for "manual" onboarding, it makes makes sense to detail this but it would go on the docs website, probably in the FAQ section.

**Comment by jessedmatlock** [comment]
I'm a little late to the party.. disregarding the sunset notice until yesterday when Atom wouldn't even open on my Mac. 
So, now I'd like to use Pulsar but don't see any clear docs on migrating settings, etc.. 
Can someone point me, and others in the same boat, to some resources - if they exist.
If they don't exist, I'm more than willing to help in any way that I can - testing, UX, etc.

**Reply by confused-Techie** [reply]
Yeah I don't think we have any docs as of yet.

But the simplest and best thing you can do to migrate your old Atom settings over is to simply copy the contents of your `.atom` folder to `.pulsar` and relaunch the editor.

That should work just fine for the vast majority if not all users. 

But it would likely be a good idea to get this properly documented somewhere.

**Reply by jessedmatlock** [reply]
Where can one find the .atom folder??

**Reply by confused-Techie** [reply]
If you've installed Atom previously, it should just be `~/.atom` or otherwise `/Users/<your_user>/.atom`

Then `.pulsar` should be in the same location just `~/.pulsar`

Keep in mind that this is a hidden folder, which Finder won't show by default. Meaning you either need to enable the option to show hidden folders within Finder or navigate to this folder via the terminal.
